sdks/java/io/kafka/src/test/java/org/apache/beam/sdk/io/kafka/KafkaIOTest.java

```diff
@@ -1065,6 +1066,55 @@ public void testValuesSink() throws Exception {
     }
   }
 
+  @Test
+  public void testRecordsSink() throws Exception {
+    // Simply read from kafka source and write to kafka sink using ProducerRecord transform. Then
+    // verify the records are correctly published to mock kafka producer.
+
+    int numElements = 1000;
+
+    try (MockProducerWrapper producerWrapper = new MockProducerWrapper()) {
+
+      ProducerSendCompletionThread completionThread =
+          new ProducerSendCompletionThread(producerWrapper.mockProducer).start();
+
+      String topic = "test";
+
+      p.apply(mkKafkaReadTransform(numElements, new ValueAsTimestampFn()).withoutMetadata())
+          .apply(ParDo.of(new KV2ProducerRecord(topic)))
+          .setCoder(ProducerRecordCoder.of(VarIntCoder.of(), VarLongCoder.of()))
```

Comment:
Do you need to set a coder here?

Reply:
I think yes, otherwise pipeline is not aware which coder to use for `ProducerRecord`'s. Can we avoid this?